src/tests/Interop/ObjectiveC/autorelease.mm

```diff
@@ -31,3 +32,10 @@ - (void)release
 {
     return NumReleaseCalls;
 }
+
+using ManagedCallback = void(*)(AutoReleaseTest*);
+extern "C" DLL_EXPORT void passAndCallOnNativeThread(AutoReleaseTest* art, ManagedCallback callback)
+{
+    std::thread th{ callback, art };
```

Comment:
I understand that this does what we agreed to, but I am wondering what it means for reliability.

The way we create the autorelease pool for the external threads violates Apple guidable: You should always drain an autorelease pool in the same context (invocation of a method or function, or body of a loop) that it was created. (From https://developer.apple.com/documentation/foundation/nsautoreleasepool.)

What happens when there is autorelease pool created on the thread already, and we push the abandoned people on top of it? i.e.

ExternalThread()
{
    3rd party ObjectiveC code

    @autoreleasepool {
        .... the 3rd party ObjectiveC code here will call into the .NET runtime that will push our new abandoned autorelease pool ...
    }

    more 3rd party Objective code
}
```

Reply:
> The way we create the autorelease pool for the external threads violates Apple guideline: You should always drain an autorelease pool in the same context (invocation of a method or function, or body of a loop) that it was created.

Agreed. Xamarin does basically what we do here minus the implicit drain. The release is only when it detaches from the runtime - at least that is how I understand the Mono Embedding APIs work.

> What happens when there is autorelease pool created on the thread already, and we push the abandoned people on top of it?

My understanding is the above `autoreleasepool` block will add a new pool prior to calling into the runtime. When it exits it would pop the one we would have added off the top of the thread's pool stack. This would then cause a pool that doesn't drain. I'm not sure how we want to handle this. Perhaps @rolfbjarne has an opinion here. I can move this around without a lot of issue.

Reply:
> > The way we create the autorelease pool for the external threads violates Apple guideline: You should always drain an autorelease pool in the same context (invocation of a method or function, or body of a loop) that it was created.
> 
> Agreed. Xamarin does basically what we do here minus the implicit drain. The release is only when it detaches from the runtime - at least that is how I understand the Mono Embedding APIs work.
> 
> > What happens when there is autorelease pool created on the thread already, and we push the abandoned people on top of it?
> 
> My understanding is the above `autoreleasepool` block will add a new pool prior to calling into the runtime. When it exits it would pop the one we would have added off the top of the thread's pool stack. This would then cause a pool that doesn't drain. I'm not sure how we want to handle this. Perhaps @rolfbjarne has an opinion here. I can move this around without a lot of issue.

I've always wanted Xamarin to follow Apple's guidelines, but it wasn't my decision, nor did our approach ever break, so I never pushed for it (and there's always broken stuff that ends up taking priority).

In my experience Apple tends to break code that doesn't follow their guidelines, so I've been kind of waiting for this to break at some point.

My opinion is that even if the current approach works today, it's preferable to follow the guidelines.

Reply:
>My opinion is that even if the current approach works today, it's preferable to follow the guidelines.

@rolfbjarne Sounds good.

The plan is therefore to update this PR to the following:

1) Newly created managed threads will get a `NSAutoreleasePool` allocated for them and then during termination of the thread we will explicitly call `drain` on the pool. We will not be relying on the statement in [doc](https://developer.apple.com/documentation/foundation/nsautoreleasepool) that states: "When a thread terminates, it automatically drains all of the autorelease pools associated with itself."

1) When a thread enters the runtime from unmanaged code a new `NSAutoreleasePool` will be allocated for the thread. When the thread returns from the Reverse P/Invoke we will call `drain` on the pool. This does mean there may be some impact to scenarios such as methods marked with `UnmanagedCallersOnly`. Optimizations could exist here to skip allocating/draining the pool when a thread that already has an runtime allocated pool enters from unmanaged code.

Reply:
I think we are going to augment the above plan of record. Step (1) is still a go, however (2) is not. The current plan is to _not_ push an `NSAutoreleasePool` when an external thread enters the runtime for the first time. I've documented our reasoning in code and will duplicate it here. In this case, the term "platform initialization" indicates pushing an `NSAutoreleasePool`.

```c++
// When a new thread enters the runtime through a Reverse P/Invoke
// skipping platform initialization is the preferred manner. The reasoning
// is based on the fact that users may have already initialized this thread
// for the platform in question. Attempting to re-initialize this thread is
// therefore imposing policy in a manner that may be undesirable.
```